### PR TITLE
[Feature] add auto purge on memory event

### PR DESCRIPTION
**Pull request by rromanchuk**
This will probably be more controversial (it shouldn't be though)  so i separated it from #1140 

Inspired from 
http://stackoverflow.com/a/19549090/775762

related to #538

### CODE REVIEW

**Comment by markrickert** [comment]
I like this, but do we want to remove ALL objects? Is there a way to remote just everything but the most recently cached _n_ objects?

**Comment by rromanchuk** [comment]
@markrickert yeah that would be ideal, but from what i can tell `NSCache` is like angsty teenager and it's eviction policies are just suggestions. I'm still uncomfortable with this because `SDWebImage` should never be causing system memory events. I found this actually can be managed with #1140  At this point the implementer has clearly mismanaged his memory and it's already too late to be algorithmic about it. 

For over a year I have been fighting a bug where a long session would run up the memory and crash/shutdown when a memory intensive action followed. I was always under the assumption that this was perfectly normal to run up this cache as the system would reclaim it when needed, i figured it was just some memory leak inside other libraries like `GPUImage` when in reality `NSCache` would drive up memory pressure and the device would send a memory warning, `NSCache` would ignore it, and the app would explode on a memory intensive alloc. 

I hate this PR, but short of completely turning off NSCache as https://github.com/rs/SDWebImage/pull/1130 or rewriting a real LRU eviction policy, we are being terrible netizens at the moment. 

A good start might be to leave a message in the README about carefully configuring SDWebImage on the context of the app and to carefully monitor memory so that maximum cost and count limit are set correctly

File: SDWebImage/SDImageCache.h [hunk 1]
```diff
@@ -48,6 +48,11 @@ typedef void(^SDWebImageCalculateSizeBlock)(NSUInteger fileCount, NSUInteger tot
 @property (assign, nonatomic) NSUInteger maxMemoryCost;
 
 /**
+ * The maximum number of objects the cache should hold.
+ */
+@property (assign, nonatomic) NSUInteger maxMemoryCountLimit;
```
Comment (rs) [inline SDWebImage/SDImageCache.h]: This was part of another PR right?

File: SDWebImage/SDImageCache.m [hunk 2]
```diff
@@ -377,6 +400,14 @@ - (NSUInteger)maxMemoryCost {
...
 }
 
+- (NSUInteger)maxMemoryCountLimit {
+    return self.memCache.countLimit;
+}
+
+- (void)setMaxMemoryCountLimit:(NSUInteger)maxCountLimit {
+    self.memCache.countLimit = maxCountLimit;
+}
+
```
Comment (rs) [inline SDWebImage/SDImageCache.m]: This chunk was part of another PR right?

**Comment by rs** [comment]
I agree, the right fix is to implement an LRU cache. I don't have the bandwidth to do that right know, but if someone want to send a PR, I have the feeling that a lot of people would be grateful.

In the meantime I'm gonna accept this PR as it is a good meantime tradeoff. Could you please remove the part of the PR that belongs to anothers?

**Comment by rromanchuk** [comment]
closing for https://github.com/rs/SDWebImage/pull/1143

**Pull request closed without merging**